### DISCUSSION TITLE
what am i missing in this onModelAfterCreate pbhook

### DISCUSSION THREAD

**Discussion by pvbhanuteja** [pocketbase/pocketbase#4167] [Q&A] [answered]
I'm trying to create a user profile api key as soon as user is created but there seems to be some error 

```
// fires only for "users" and "members"
onModelAfterCreate((e) => {

    // id , created , updated , username , email , emailVisibility , verified 
    console.log(e.model.tableName())
    console.log(e.model.id)
    // Find the collection by name
    const postsCollection = $app.dao().findCollectionByNameOrId('user_profile')
    console.log(postsCollection)
    // Create a new post record
    const record = new Record(postsCollection, {
        user: e.model.id,
        plan: 'free',
        sub_id: null,
        apiKey: generateUUIDTypeAPIKey(),
        usage: {
            checks1: 0,
            checks2: 0,
            checks3: 0
        },
        next_period: addMonthToCurrentTimestamp()
    })
    console.log('created',record)
    // Save the record
    $app.dao().saveRecord(record)
}, "users", "members")
```

Also can we do something like this ?

```
onModelBeforeCreate((e) => {

// get the user.id for relation.

Continue with the user creation logic


// Then crate user_profile with api key 
}
```

The second logic is more reliable as we exactly know the user and user_profile will be created on the same request.

**Comment by ganigeorgiev** [comment]
> but there seems to be some error

Could you paste the error?

**Reply by pvbhanuteja** [reply]
Actually I can't find error

2024/01/17 05:32:21 Server started at http://127.0.0.1:8090
├─ REST API: http://127.0.0.1:8090/api/
└─ Admin UI: http://127.0.0.1:8090/_/
2024/01/17 05:32:41 users
2024/01/17 05:32:41 hsvwcvx15jyhd2j
2024/01/17 05:32:41 [object Object]

I logged 4 statements but the 4th statement isn't executed.

**Reply by ganigeorgiev** [reply]
Try to start the executable with `--dev` and/or change the "Min log level" to `-4` in the _Admin UI > Logs > Logs settings_ popup.

**Reply by pvbhanuteja** [reply]
```
[0.00ms] INSERT INTO `_externalAuths` (`collectionId`, `created`, `id`, `provider`, `providerId`, `recordId`, `updated`) VALUES ('_pb_users_auth_', '2024-01-17 14:01:13.199Z', 'k8wgjryfomu04eh', 'github', '33227278', 'eczv1jzxcu9lijk', '2024-01-17 14:01:13.199Z')
2024/01/17 06:01:13 users
2024/01/17 06:01:13 eczv1jzxcu9lijk
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = 'user_profile' OR LOWER([[name]])='user_profile' LIMIT 1
2024/01/17 06:01:13 [object Object]
ERROR POST /api/collections/users/auth-with-oauth2
└─ Failed to authenticate.
   └─ after transaction errors: ReferenceError: generateUUIDTypeAPIKey is not defined at <eval>:14:39(40)
DEBUG Realtime connection closed (cancelled request)
└─ map[clientId:JaoHD6GuQ6wiR2XaYGkUB70KiCVP1R5yeIdyuRTO]
```

Should the function definitions be inside onModelAfterCreate? I casually logged them they worked fine.

**Reply by ganigeorgiev** [reply, marked as answer]
If your functions `generateUUIDTypeAPIKey` and `addMonthToCurrentTimestamp` are global then this will not work because of the [handlers scope limitation](https://pocketbase.io/docs/js-overview/#handlers-scope).

You'll have to define them inside the hook or move them in a separate js file and load them with `require`.

**Reply by pvbhanuteja** [reply]
Got it. I also tried doing the below. How can I make sure the model gets created? I want to have that user model id for the creation of user_profile is it achievable by onModelBeforeCreate ?

```
onModelBeforeCreate((e) => {

// get the user.id for relation.

Continue with the user creation logic(default pb user creation)


// Then crate user_profile with api key 
}
```

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand what is the desired behavior.
The after hook guarantees that your code will be executed **after** the model is created.
There is no such guarantee for the before hook because the model persistence can fail (eg. because of a unique db constraint error).

**Reply by pvbhanuteja** [reply]
The problem with onModelAfterCreate is I'm worried that what if user_profile related API is called as soon as user signup/creation is done. I'm not sure of the delay for onModelAfterCreate. Is this hook run after the API return with 200 responses or will this run before hand for user creation?

**Reply by ganigeorgiev** [reply]
> Is this hook run after the API return with 200 responses

No. The `onModelAfterCreate` will be fired before returning the response and its execution is synchronous/blocking for its related request goroutine.

**Reply by ganigeorgiev** [reply]
If you are trying to execute the user create operation in a transaction with your custom hook code, then this is not possible at the moment.

**Comment by jonshipman** [comment]
I have better luck running dao operations in hooks (either builder or newQuery). See https://pocketbase.io/docs/js-database/#query-builder.

Your generateUUIDTypeAPIKey and addMonthToCurrentTimestamp need to be included in the top of your function. 

```
const file = $filepath.join(__hooks, 'utils.js');
const { generateUUIDTypeAPIKey, addMonthToCurrentTimestamp } = require(file);
```
(assumes the functions are in a file named "utils.js" in your pb_hooks folder)

Like @ganigeorgiev said.
